datafusion/core/src/physical_optimizer/join_selection.rs

```diff
@@ -1287,6 +1289,29 @@ mod tests_statistical {
         );
     }
 
+    #[tokio::test]
+    async fn test_hash_join_swap_on_joins_with_projections() -> Result<()> {
+        let (big, small) = create_big_and_small();
+        let join = Arc::new(HashJoinExec::try_new(
+            Arc::clone(&big),
+            Arc::clone(&small),
+            vec![(
+                Arc::new(Column::new_with_schema("big_col", &big.schema())?),
+                Arc::new(Column::new_with_schema("small_col", &small.schema())?),
+            )],
+            None,
+            &JoinType::Inner,
+            Some(vec![1]),
+            PartitionMode::Partitioned,
+            false,
+        )?);
+        let swapped = swap_hash_join(&join.clone(), PartitionMode::Partitioned)
+            .expect("swap_hash_join must support joins with projections");
+        assert_eq!(swapped.schema().fields.len(), 1);
+        assert_eq!(swapped.schema().fields[0].name(), "small_col");
+        Ok(())
+    }
+
```

Comment:
I am not sure whether this asserts test a projection exists or not 🤔

Reply:
I suppose [checking](https://github.com/apache/datafusion/pull/12967/files#diff-4708a2cf52c334b07864e98889c4b818126b846bd6551fe1a467be17f3d72f4cR1310) that output schema of swapped join contains one field (which is what's been passed as a projection arg) should be enough to assert that projection is not lost after swap.

Reply:
> I am not sure whether this asserts test a projection exists or not 🤔

`swapped` is `HashJoinExec` here.  Maybe be better to add `assert_eq!(swapped.projection, Some(vec![0]))` here?

Reply:
I meant these assertions do not know if a projection is added or not. Isn't it  better to assert a new condition like `assert!(swapped.as_any().is::<HashJoinExec>());` since 
```
assert_eq!(swapped.schema().fields.len(), 1);
assert_eq!(swapped.schema().fields[0].name(), "small_col");
```
 those ones can be also satisfied by a `ProjectionExec`. Am I missing smth?

Reply:
good idea.

Reply:
done

Reply:
> these assertions do not know if a projection is added or not

Got it -- I was thinking only about embedded projection, and completely missed that it's also worth checking that no additional projection is added on top of join :+1: